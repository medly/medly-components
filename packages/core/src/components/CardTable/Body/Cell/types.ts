import { ColumnConfig, Data } from '../../types';

export interface Props {
    data: any;
    column: ColumnConfig;
    rowId: any;
    isRowClickDisabled?: boolean;
    rowData: Data;
}

export interface StyledProps {
    align?: 'left' | 'center' | 'right';
    wrapText?: boolean;
}
