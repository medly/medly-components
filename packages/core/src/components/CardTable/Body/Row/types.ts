import { WithThemeProp } from '@medly-components/utils';

export interface Props extends WithThemeProp {
    onClick?: () => void;
    isRowClickDisabled?: boolean;
    withWhiteBackground?: boolean;
    gridTemplateColumns: string;
}
