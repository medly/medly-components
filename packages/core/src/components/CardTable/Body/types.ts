import { ColumnConfig, Data } from '../types';

export interface Props {
    data: Data[];
    columns: ColumnConfig[];
    onRowClick?: (rowData: Record<string, unknown>) => void;
    rowClickDisableKey?: string;
    uniqueKeyName?: string;
    withWhiteBackground?: boolean;
}
