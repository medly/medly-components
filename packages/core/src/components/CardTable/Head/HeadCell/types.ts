import { ColumnConfig, SortOrder } from '../../types';

export interface Props {
    defaultSortOrder?: SortOrder | 'none';
    column: ColumnConfig;
    sortField?: string;
    withWhiteBackground?: boolean;
    onSortChange?: (field: string, order: SortOrder) => void;
}

export interface StyledProps {
    isActive?: boolean;
    sortable?: boolean;
    withWhiteBackground?: boolean;
    align?: 'left' | 'center' | 'right';
}
