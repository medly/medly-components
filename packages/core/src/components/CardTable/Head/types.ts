import { ColumnConfig, SortOrder } from '../types';

export interface Props {
    defaultSortField?: string;
    defaultSortOrder?: SortOrder;
    withWhiteBackground?: boolean;
    columns: ColumnConfig[];
    onSort?: (field: string, order: SortOrder) => void;
}
