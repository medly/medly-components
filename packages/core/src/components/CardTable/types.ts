import { HTMLProps } from '@medly-components/utils';
import type { FC } from 'react';

export type SortOrder = 'asc' | 'desc';

export interface Data {
    [key: string]: any;
}

export interface ColumnConfig {
    /** Title of the column */
    title: string;
    /** Field name in the data */
    field: string;
    /** Set it true to show sort icons */
    sortable?: boolean;
    /** Column content alignment */
    align?: 'left' | 'right' | 'center';
    /** Column Text content wrapping */
    wrapText?: boolean;
    /** Provide the fraction of the of the total width of the row */
    fraction?: number;
    /** Custom component */
    component?: FC<{ data: any; rowId?: any; disabled?: boolean; rowData?: Data }>;
}
export interface CardTableProps extends Omit<HTMLProps<HTMLOListElement>, 'data' | 'type'> {
    /** Table data */
    data: Array<{
        [key: string]: any;
    }>;
    /** Column configuration */
    columns: ColumnConfig[];
    /** Key name in data to be used as unique id */
    uniqueKeyName?: string;
    /** Default Sort Field*/
    defaultSortField?: string;
    /** Default Sort Order*/
    defaultSortOrder?: SortOrder;
    /** Function to be called on click of sort icon */
    onSort?: (field: string, order: SortOrder) => void;
    /** Function to be called on row click */
    onRowClick?: (rowData: Record<string, unknown>) => void;
    /** Key name to disable row click */
    rowClickDisableKey?: string;
    /** Key name to select background color */
    withWhiteBackground?: boolean;
}
