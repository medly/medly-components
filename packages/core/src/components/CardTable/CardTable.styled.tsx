import styled from 'styled-components';

export const Table = styled('table')`
    width: 100%;
`;
