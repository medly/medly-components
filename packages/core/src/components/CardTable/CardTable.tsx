import React, { FC } from 'react';
import Body from './Body';
import * as Styled from './CardTable.styled';
import Head from './Head';
import { CardTableProps } from './types';

export const CardTable: FC<CardTableProps> = React.memo(
    ({ withWhiteBackground, defaultSortField, defaultSortOrder, onSort, columns, ...restProps }) => (
        <Styled.Table>
            <Head {...{ defaultSortField, defaultSortOrder, onSort, columns, withWhiteBackground }} />
            <Body {...{ columns, withWhiteBackground, ...restProps }} />
        </Styled.Table>
    )
);

CardTable.displayName = 'CardTable';
CardTable.defaultProps = {
    withWhiteBackground: false
};
