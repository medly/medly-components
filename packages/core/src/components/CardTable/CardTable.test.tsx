import { fireEvent, render } from '@test-utils';
import { CardTable } from './CardTable';
import { ColumnConfig } from './types';

const transactions = [
        {
            id: 1,
            patientInfo: 'Oli Bob',
            rxInfo: 'Metaformin',
            price: '535.76',
            status: 'Pending'
        }
    ],
    transactionColumnConfig: ColumnConfig[] = [
        { title: 'Patient Info', field: 'patientInfo', sortable: true },
        { title: 'RxInfo', field: 'rxInfo', sortable: true }
    ],
    renderer = ({
        onSort = jest.fn(),
        onRowClick = jest.fn(),
        withWhiteBackground = false,
        data = transactions,
        columns = transactionColumnConfig
    }) =>
        render(
            <CardTable
                data={data}
                onSort={onSort}
                columns={columns}
                onRowClick={onRowClick}
                withWhiteBackground={withWhiteBackground}
                rowClickDisableKey="disable"
                uniqueKeyName="123"
            />
        );

describe('CardTable', () => {
    it('should render properly with grey background', () => {
        const { container } = renderer({ withWhiteBackground: false });
        expect(container).toMatchSnapshot();
    });
    it('should render properly with white background', () => {
        const { container } = renderer({ withWhiteBackground: true });
        expect(container).toMatchSnapshot();
    });
    it('should call onSort with expected data', () => {
        const onSort = jest.fn(),
            { container } = renderer({ withWhiteBackground: true, onSort });
        fireEvent.click(container.querySelectorAll('svg')[0]);
        expect(onSort).toHaveBeenCalledWith('patientInfo', 'asc');
    });
    it('should call onRowClick on click on any row', () => {
        const onRowClick = jest.fn(),
            { getByText } = renderer({ withWhiteBackground: true, onRowClick });
        fireEvent.click(getByText('Oli Bob'));
        expect(onRowClick).toHaveBeenCalled();
    });
});
