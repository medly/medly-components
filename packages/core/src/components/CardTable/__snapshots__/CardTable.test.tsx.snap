// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTable should render properly with grey background 1`] = `
.c6 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: initial;
  text-transform: uppercase;
}

.c11 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c10 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

.c10 .c5 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c9 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  cursor: pointer;
  box-sizing: border-box;
  border-radius: 0.8rem;
  margin-bottom: 0.8rem;
  background-color: white;
  -webkit-transition: box-shadow 100ms ease-out;
  transition: box-shadow 100ms ease-out;
  border: 0.1rem solid #ffffff;
  box-shadow: 0 0.2rem 0.8rem rgba(176,188,200,0.2);
}

.c9:hover {
  box-shadow: 0 0.4rem 0.8rem rgba(176,188,200,0.6);
}

.c9:nth-child(odd) {
  background: #ffffff;
}

.c9:nth-child(even) {
  background: #ffffff;
}

.c0 {
  width: 100%;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background-color: transparent;
}

.c8 {
  overflow: visible;
  font-size: 2rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c8 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c4 {
  padding: 1rem 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  padding: 0 0.8rem;
  margin: 1.4rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2::after {
  content: '';
  width: calc(100% - 1.6rem);
  margin: 0 0.8rem;
  display: block;
  position: absolute;
  bottom: -1.6rem;
  left: 0;
  background-color: transparent;
  height: 0.2rem;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.2rem 0.2rem 0 0;
}

.c2 .c3:hover {
  cursor: pointer;
}

.c2 .c3:hover .c5 {
  color: #5687D9;
}

.c2 .c3:hover .c7 * {
  fill: #5687D9;
}

.c2 .c5 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline;
  color: #435465;
}

.c2 .c7 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  width: 2.4rem;
  cursor: pointer;
  padding-left: 0.4rem;
}

.c2 .c7 * {
  fill: #435465;
}

<div>
  <table
    class="c0"
  >
    <thead>
      <tr
        class="c1"
      >
        <th
          class="c2"
        >
          <span
            class="c3 c4"
          >
            <h5
              class="c5 c6"
            >
              Patient Info
            </h5>
            <svg
              class="c7 c8"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
        <th
          class="c2"
        >
          <span
            class="c3 c4"
          >
            <h5
              class="c5 c6"
            >
              RxInfo
            </h5>
            <svg
              class="c7 c8"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="c9"
      >
        <td
          class="c10"
        >
          <span
            class="c5 c11"
            title="Oli Bob"
          >
            Oli Bob
          </span>
        </td>
        <td
          class="c10"
        >
          <span
            class="c5 c11"
            title="Metaformin"
          >
            Metaformin
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`CardTable should render properly with white background 1`] = `
.c6 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: initial;
  text-transform: uppercase;
}

.c11 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c10 {
  padding: 0;
  padding: 1.5rem 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  overflow: hidden;
}

.c10 .c5 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c9 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  cursor: pointer;
  box-sizing: border-box;
  border-radius: 0.8rem;
  margin-bottom: 0.8rem;
  background-color: white;
  -webkit-transition: box-shadow 100ms ease-out;
  transition: box-shadow 100ms ease-out;
  border: 0.1rem solid #dfe4e9;
  box-shadow: none;
}

.c9:hover {
  box-shadow: 0px 0.2rem 0.4rem rgba(176,188,200,0.6);
}

.c9:nth-child(odd) {
  background: #F7F8F9;
}

.c9:nth-child(even) {
  background: #ffffff;
}

.c0 {
  width: 100%;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background-color: transparent;
}

.c8 {
  overflow: visible;
  font-size: 2rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c8 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c4 {
  padding: 1rem 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  padding: 0 0.8rem;
  margin: 1.4rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2::after {
  content: '';
  width: calc(100% - 1.6rem);
  margin: 0 0.8rem;
  display: block;
  position: absolute;
  bottom: -1.6rem;
  left: 0;
  background-color: transparent;
  height: 0.2rem;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.2rem 0.2rem 0 0;
}

.c2 .c3:hover {
  cursor: pointer;
}

.c2 .c3:hover .c5 {
  color: #5687D9;
}

.c2 .c3:hover .c7 * {
  fill: #5687D9;
}

.c2 .c5 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline;
  color: #607890;
}

.c2 .c7 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  width: 2.4rem;
  cursor: pointer;
  padding-left: 0.4rem;
}

.c2 .c7 * {
  fill: #607890;
}

<div>
  <table
    class="c0"
  >
    <thead>
      <tr
        class="c1"
      >
        <th
          class="c2"
        >
          <span
            class="c3 c4"
          >
            <h5
              class="c5 c6"
            >
              Patient Info
            </h5>
            <svg
              class="c7 c8"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
        <th
          class="c2"
        >
          <span
            class="c3 c4"
          >
            <h5
              class="c5 c6"
            >
              RxInfo
            </h5>
            <svg
              class="c7 c8"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="c9"
      >
        <td
          class="c10"
        >
          <span
            class="c5 c11"
            title="Oli Bob"
          >
            Oli Bob
          </span>
        </td>
        <td
          class="c10"
        >
          <span
            class="c5 c11"
            title="Metaformin"
          >
            Metaformin
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
